### ISSUE TITLE
gpfdist-sink using incorrect address to serve data

### ISSUE THREAD

**Issue by sabbyanandan** [spring-cloud/spring-cloud-stream-app-starters#77]
_From @jpatel-pivotal on February 9, 2016 23:45_

I have deployed the scdf gpfdist module to PCF. I also have GPDB running in the same IAAS. 

The spring-boot app (module)  in this case can connect to my gpdb environment and create the external table it needs as shown below:

```
postgres=# \d
                                    List of relations
 Schema |                       Name                        | Type  |  Owner  | Storage
--------+---------------------------------------------------+-------+---------+----------
 public | ticktock                                          | table | gpadmin | heap
 public | ticktock_ext_30526455_9ae8_45aa_854b_43ea3897a3b3 | table | gpadmin | external
 public | ticktock_ext_41debee3_2222_45a9_bcc5_16e5f35e9369 | table | gpadmin | external
 public | ticktock_ext_f4e17689_ab0c_45fd_b30a_d91c78438641 | table | gpadmin | external
```

However if you look at the table definition for the external table, it is using the container's IP as shown below:

```
postgres=# \d+ ticktock_ext_f4e17689_ab0c_45fd_b30a_d91c78438641
External table "public.ticktock_ext_f4e17689_ab0c_45fd_b30a_d91c78438641"
 Column | Type | Modifiers | Storage  | Description
--------+------+-----------+----------+-------------
 date   | text |           | extended |
 time   | text |           | extended |
Type: readable
Encoding: UTF8
Format type: text
Format options: delimiter ' ' null '\N' escape '\'
External location: gpfdist://10.254.0.14:8080/data
```

The external location is using the container's ip address derived using httpServer.getListenAddress() which changes each time the container is restarted or if new container instances spawn up, etc. That network is an arbitrary network and is not routable unless you go out and come back in using the route that you get provided when you deploy the app in PCF. 

So in short the external connection to GPDB to create the external table works but when you want to read data from that external table, the app logs a lot of errors that look like below because it can not connect to the container that is running gpfdist and serving out the data:

```
2016-02-09T23:40:12.000+00:00 [APP] OUT Caused by: org.postgresql.util.PSQLException: ERROR: connection with gpfdist failed for gpfdist://10.254.0.14:8080/data. effective url: http://10.254.0.14:8080/data. error code = 110 (Connection timed out) (seg3 slice1 sdw1:40003 pid=9758)
```

_Copied from original issue: spring-cloud/spring-cloud-stream-modules#182_

**Comment by sabbyanandan** [comment]
_From @jpatel-pivotal on February 10, 2016 1:13_

Here is how I got it to work:
```
curl -vvv "http://router_status:<PASSWORD>@<ROUTER_IP>:8080/routes"
```
That gave me the address:port info that I needed to use in the location URI when creating a table and then it all worked. 

Below is what my test table DDL looked like:

```
postgres=# \d+ test
            External table "public.test"
 Column | Type | Modifiers | Storage  | Description
--------+------+-----------+----------+-------------
 date   | text |           | extended |
 time   | text |           | extended |
Type: readable
Encoding: UTF8
Format type: text
Format options: delimiter ' ' null 'N' escape '\'
External location: gpfdist://10.65.96.33:60034/data
```

The above external location that the PCF Router gave me was routable and so I was able to select * from test correctly.

However I don't know what happens when the route changes in case the container gets re-spawned somewhere else or the case where you may want to scale multiple instances of it? So not saying that is how you should fix it :)
